### ISSUE TITLE
Functions in class cannot be analyzed under some circumstances

### ISSUE THREAD

**Issue by BradXiao** [microsoft/pylance-release#2745]
<h3> Environment data </h3>

-   Language Server version: 2022.4.3
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): Anaconda Python 3.7.7
-   python.analysis.indexing: null
-   python.analysis.typeCheckingMode: off

<h3> Expected behaviour </h3>

![圖片](https://user-images.githubusercontent.com/20597547/165686459-dd5a1311-f22d-4d87-a7af-06ad785f1f6c.png)

<h3> Actual behaviour </h3>

![圖片](https://user-images.githubusercontent.com/20597547/165686090-279ea185-87ad-4557-8822-4c49a0ab273c.png)

<h3> Code Snippet / Additional information </h3>

```python
import random
class apple():
    rot_way:str

    def __init__(self) -> None:
        self.rot_time='3 days'
    def rot(self, *args):
        print(f'will rot in {self.rot_time}')
        return random.choise(['ok','eaten by bugs'])
    def get_rot_way(self):
        return self.rot_way      
class grow():
    location:str
    spped:str
def func():
    fruit=apple()

    grow_recipe:grow=None

    if grow_recipe is None:
        if fruit.rot()=='ok':
            grow_recipe=grow_recipe()
            grow_recipe.location='Lin Garden'
            grow_recipe.spped='3 days'

    if grow_recipe is not None:
        print(f'apple will reborn at {grow_recipe.location}')
        return

    idx=[i for i in [] if i=='dummy']


    fruit.get_rot_way()
```

**Comment by erictraut** [comment]
The problem here is that you have declared the type of the variable `grow_recipe` to be `grow`, but you have assigned it a value of `None`, which is not compatible. Later in the code, you test whether `grow_recipe is not None`. The type analyzer concludes that this will always be true because `grow_recipe` is declared to be an object of type `grow`. It therefore concludes that `func` will always execute this code block and return.

You can correct the type declaration for `grow_recipe` by doing the following:

```python
    grow_recipe: grow | None = None
```

Or if you are using Python 3.9 or earlier:
```python
    grow_recipe: Optional[grow] = None
```

Since you are using type annotations, I recommend enabling type checking diagnostics by setting "python.analysis.typeCheckingMode" to "basic". By default, type checking diagnostics are disabled in pylance.

**Comment by erictraut** [comment]
There is something odd here though. My diagnosis above doesn't entirely explain the symptoms. If I comment out the line starting with `idx = ...`, then `fruit.get_rot_way()` no longer evaluates to type `Never`. I'll take a closer look at this. Since this is a core type checking issue, I've opened a [tracking bug pyright project](https://github.com/microsoft/pyright/issues/3395).

**Comment by erictraut** [comment]
There was a bug that led to an inconsistency in type evaluation in this case. I've addressed the inconsistency, and the type of `fruit` now evaluates to the type `apple` regardless of whether the statement `idx = ...` is present or not. This will be addressed in the next release of pylance.

**Comment by BradXiao** [comment]
> There was a bug that led to an inconsistency in type evaluation in this case. I've addressed the inconsistency, and the type of `fruit` now evaluates to the type `apple` regardless of whether the statement `idx = ...` is present or not. This will be addressed in the next release of pylance.

Thank you very much for the quick replying and the fix.

**Comment by debonte** [comment]
@BradXiao, the fix is available in the latest Pylance insiders build ([2022.5.0-pre.1](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202250-pre1-4-may-2022)). We're not shipping a stable build this week. If you're not using Pylance insiders, you'll see the fix in next week's stable build.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.5.1, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202251-11-May-2022)
